nio/src/main/scala/zio/nio/channels/AsynchronousChannel.scala

```diff
@@ -151,10 +139,10 @@ object AsynchronousByteChannel {
   private[channels] def effectAsyncChannel[C <: JChannel, A](
     channel: C
   )(op: C => CompletionHandler[A, Any] => Any): IO[IOException, A] =
-    IO.effectAsyncInterrupt { k =>
-      op(channel)(completionHandlerCallback(k))
-      Left(IO.effect(channel.close()).ignore)
-    }
+    Task(op(channel))
+      .flatMap(Task.effectAsyncWithCompletionHandler)
+      .refineToOrDie[IOException]
+      .onInterrupt(IO.effect(channel.close()).ignore)
```

Comment:
I'm guessing here, but wouldn't `onInterrupt` be enough here, @quelgar ? :thinking:

Reply:
Yeah that sounds like it would work.